guidance/library/_json.py

```diff
@@ -92,12 +92,12 @@ def validate_json_node_keys(node: Mapping[str, Any]):
         )
 
 
-@guidance(stateless=True)
+@guidance(stateless=True, cache=True)
```

Comment:
If adding cache here gives a speed up, will this result in every generated integer being the same?

Reply:
Related question, assuming that the return value is OK: does this get cross-referenced properly when the grammar is generated? Looking at some of the serialisation code, I expect that it does, but would be good to make sure that it doesn't get expanded inline multiple times.

Reply:
`cache=True` will ensure that subsequent calls with the same arguments will always return the exact same object (i.e. `id(o1) == id(o2)`).

The serialization code maintains its own cache for handling "pointers" (just integer indices) to objects in the serialized grammar, as you can see [here](https://github.com/guidance-ai/guidance/blob/958145ccce5d23f72fec4e2c620f1a6a6df1358f/guidance/_grammar.py#L1047). `GrammarFunction`s are hashable and just hash to the `id` of the object, so this code should correctly respect the object caching we do in the decorator.